File: java/CannibalNumbers.java
import java.util.Arrays;
import java.util.Scanner;

public class CannibalNumbers {

	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		System.out.println("Enter the sacrifices: ");
		int[] integers = createIntArray(in.nextLine().split(" "));
		
		System.out.println("Enter the query: ");
		int query = Integer.parseInt(in.nextLine());
		in.close();
		
		int cannibals = findCannibals(integers.clone(), query);
		System.out.print("With the numbers " + Arrays.toString(integers));
		System.out.print(" there are " + cannibals + " cannibals");
		
	}
	
	public static int findCannibals(int[] a, int q) {
		// sort array to place all possible cannibals at the end of the array
		// and all the numbers that will be consumed at the beginning of the array
		int cannibals = 0;
		Arrays.sort(a);
		int nextSacrifice = 0;
		int nextCannibal = a.length - 1;
		
		// if next number to be eaten is the number that will be doing the eating then there is no one else to eat
		while(nextSacrifice != nextCannibal){
			if(a[nextCannibal] >= q) {
				cannibals++;
				nextCannibal--;
			}
			else {
				a[nextCannibal] += 1;
				nextSacrifice++;
			}
		}
		
		// last number doesn't get evaluated 
		if(a[nextCannibal] >= q) {
			cannibals++;
		}
		
		return cannibals;
	}
	
	public static int[] createIntArray(String[] a) {
		int[] array = new int[a.length];
		for(int i=0; i < a.length; i++) {
			array[i] = Integer.parseInt(a[i]);
		}
		return array;
	}
}

File: ruby/cannibal_numbers.rb
=begin
src => https://www.reddit.com/r/dailyprogrammer/comments/76qk58/20171016_challenge_336_easy_cannibal_numbers/

Imagine a given set of numbers wherein some are cannibals. We define a cannibal as a larger number can eat a smaller number and increase its value by 1. There are no restrictions on how many numbers any given number can consume. A number which has been consumed is no longer available.

Your task is to determine the number of numbers which can have a value equal to or greater than a specified value. 

=end

def get_cannibals(numbers, q)
	# assuming the numbers are sorted
	# all the cannibals will be at the end of the array
	# and all the sacrifices will be at the beginning of the array
	cannibals = 0
	next_cannibal = numbers.length - 1
	next_sacrifice = 0

	while next_cannibal > next_sacrifice
		if numbers[next_cannibal] >= q
			next_cannibal -= 1
			cannibals += 1
		else
			numbers[next_cannibal] += 1
			next_sacrifice += 1
		end
	end

	# last number does not get valuated as while loop exits prematurelty
	if numbers[next_cannibal] >= q then cannibals += 1 end

	return cannibals
end

if __FILE__ == $0
	numbers = "21 9 5 8 10 1 3".split.map{|i| i.to_i}.sort
	queries = 15, 10
	queries.each do |q|
		puts "For the query #{q} there are #{get_cannibals(numbers, q)} cannibals"
	end
end